src/note.rs

```diff
@@ -0,0 +1,394 @@
+                "encrypted_data",
+                &format_args!("{:?}", &self.encrypted_data),
+            )
+            .finish()
+    }
+}
+
+impl Read for Note {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let mut buf = io::BufWriter::new(&mut buf[..]);
+        let mut n = 0;
+
+        n += buf.write(&[self.note_type as u8])?;
+        n += buf.write(
+            &JubJubAffine::from(&self.value_commitment).to_bytes()[..],
+        )?;
+        n += buf.write(&self.nonce.to_bytes())?;
+        n += buf.write(&self.stealth_address.to_bytes())?;
+        n += buf.write(&self.pos.to_le_bytes())?;
+        n += buf.write(&self.encrypted_data[..])?;
+
+        buf.flush()?;
+        Ok(n)
+    }
+}
+
+impl Write for Note {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let mut buf = io::BufReader::new(&buf[..]);
+
+        let mut one_byte = [0u8; 1];
+        let mut one_scalar = [0u8; 32];
+        let mut one_u64 = [0u8; 8];
+        let mut one_stealth_address = [0u8; 64];
+
+        let mut n = 0;
+
+        buf.read_exact(&mut one_byte)?;
+        n += one_byte.len();
+        self.note_type = one_byte[0].try_into()?;
+
+        buf.read_exact(&mut one_scalar)?;
+        n += one_scalar.len();
+        self.value_commitment =
+            JubJubExtended::from(jubjub_decode::<JubJubAffine>(&one_scalar)?);
+
+        buf.read_exact(&mut one_scalar)?;
+        n += one_scalar.len();
+        self.nonce = jubjub_decode::<JubJubScalar>(&one_scalar)?;
+
+        buf.read_exact(&mut one_stealth_address)?;
+        n += one_stealth_address.len();
+        self.stealth_address =
+            StealthAddress::from_bytes(&one_stealth_address)?;
+
+        buf.read_exact(&mut one_u64)?;
+        n += one_u64.len();
+        self.pos = u64::from_le_bytes(one_u64);
+
+        buf.read_exact(&mut self.encrypted_data)?;
+        n += self.encrypted_data.len();
+
+        Ok(n)
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+impl Note {
+    /// Creates a new phoenix output note
+    pub fn new(note_type: NoteType, psk: &PublicSpendKey, value: u64) -> Self {
+        let r = JubJubScalar::random(&mut rand::thread_rng());
+        let nonce = JubJubScalar::random(&mut rand::thread_rng());
+        let blinding_factor = JubJubScalar::random(&mut rand::thread_rng());
+
+        Self::deterministic(note_type, &r, nonce, psk, value, blinding_factor)
+    }
+
+    /// Creates a new transparent note
+    pub fn transparent(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Transparent, psk, value)
+    }
+
+    /// Creates a new obfuscated note
+    pub fn obfuscated(psk: &PublicSpendKey, value: u64) -> Self {
+        Self::new(NoteType::Obfuscated, psk, value)
+    }
+
+    /// Create a new phoenix output note without inner randomness
+    pub fn deterministic(
+        note_type: NoteType,
+        r: &JubJubScalar,
+        nonce: JubJubScalar,
```

Comment:
If that's the `PoseidonCipher` nonce. The type is not correct. It should be `BlsScalar` as per https://github.com/dusk-network/Poseidon252/blob/17329418740028b571752e936b190a1f15f94c5a/src/cipher/cipher.rs#L110

Reply:
`phoenix-core` doesn't deal with `BlsScalar` directly (in the type structures I mean) but with JubJub's instead. The fact that Poseidon only deals with `BlsScalar` is an implementation details – meaning: from the point of the phoenix-core's consumer, Poseidon is a black box.

Having Nonce as BlsScalar in a public struct is an anomaly in the design, and leaking an internal type.

I talked with Togh and he doesn't care if it's JubJub or BlsScalar, and he's going to update the docs.

It's like the `value`, basically, that is passed and stored as u64 but then is converted internally as `BlsScalar: the consumers doesn't need to know that, they shouldn't care about that.

The alternative would be have the Nonce as simple sequence of 32 bytes (that "represent" a BlsScalar), from a design point of view is also valid.

Reply:
No, I'm happy then to have it as a jubjub or bls Scalar.